consensus/safety-rules/src/t_safety_rules.rs

```diff
@@ -7,13 +7,20 @@ use consensus_types::{
     vote_proposal::VoteProposal,
 };
 use libra_crypto::ed25519::Ed25519Signature;
+use libra_types::validator_change::ValidatorChangeProof;
 
 /// Interface for SafetyRules
 pub trait TSafetyRules<T> {
     /// Provides the internal state of SafetyRules for monitoring / debugging purposes. This does
     /// not include sensitive data like private keys.
     fn consensus_state(&mut self) -> Result<ConsensusState, Error>;
 
+    /// Initialize SafetyRules using an Epoch ending LedgerInfo, this should map to what was
+    /// provided in consensus_state. It will be used to initialize the ValidatorSet.
+    /// This uses a ValidatorChangeProof because there's a possibility that consensus migrated to a
+    /// new epoch but SafetyRules did not.
+    fn initialize(&mut self, proof: &ValidatorChangeProof) -> Result<(), Error>;
```

Comment:
it probably should return the waypoint/latest epoch info so that consensus can send it correct proof

Reply:
we get that through consensus_state, unless you're suggesting that we have the error return the waypoint as well, but I don't see us taking advantage of that in the near term.